lib/vrl/stdlib/src/decode_q.rs

```diff
@@ -0,0 +1,226 @@
+        nom::Err::Incomplete(_) => e.to_string(),
+    })?;
+    let mut decoded = decoded?;
+
+    // Add remaining input to the decoded string.
+    decoded.push_str(remaining);
+
+    Ok(decoded.into())
+}
+
+/// Parses input into (head, (charset, encoding, encoded text))
+fn parse_delimited_q<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
+    input: &'a str,
+) -> IResult<&'a str, (&'a str, Option<EncodedWord<'a>>), E> {
+    pair(
+        take_until("=?"),
+        opt(delimited(tag("=?"), parse_internal_q, tag("?="))),
+    )(input)
+}
+
+/// Parses inside of encoded word into (charset, encoding, encoded text)
+fn parse_internal_q<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
+    input: &'a str,
+) -> IResult<&'a str, EncodedWord<'a>, E> {
+    map(
+        separated_pair(
+            opt(take_until1("?")),
+            tag("?"),
+            separated_pair(
+                take_until("?"),
+                tag("?"),
+                alt((take_until("?="), |input| Ok(("", input)))),
+            ),
+        ),
+        |(charset, (encoding, input))| EncodedWord {
+            charset,
+            encoding,
+            input,
+        },
+    )(input)
+}
+
+struct EncodedWord<'a> {
+    charset: Option<&'a str>,
+    encoding: &'a str,
+    input: &'a str,
+}
+
+impl<'a> EncodedWord<'a> {
+    fn decode_word(&self) -> Result<String> {
+        // Modified version from https://github.com/staktrace/mailparse/blob/a83d961fe53fd6504d75ee951a0e91dfea03c830/src/header.rs#L39
+
+        // Decode
+        let decoded = match self.encoding {
+            "B" | "b" => BASE64_MIME
+                .decode(self.input.as_bytes())
+                .map_err(|_| "Unable to decode base64 value")?,
+            "Q" | "q" => {
+                // The quoted_printable module does a trim_end on the input, so if
+                // that affects the output we should save and restore the trailing
+                // whitespace
+                let to_decode = self.input.replace('_', " ");
+                let trimmed = to_decode.trim_end();
+                let mut d = quoted_printable::decode(&trimmed, quoted_printable::ParseMode::Robust);
+                if d.is_ok() && to_decode.len() != trimmed.len() {
+                    d.as_mut()
+                        .unwrap()
+                        .extend_from_slice(to_decode[trimmed.len()..].as_bytes());
+                }
+                d.map_err(|_| "Unable to decode quoted_printable value")?
+            }
+            _ => return Err(format!("Invalid encoding: {:?}", self.encoding).into()),
+        };
+
+        // Convert to UTF-8
+        let charset = self.charset.unwrap_or("utf-8");
+        let charset = Charset::for_label_no_replacement(charset.as_bytes())
+            .ok_or_else(|| format!("Unable to decode {:?} value", charset))?;
+        let (cow, _) = charset.decode_without_bom_handling(&decoded);
+        Ok(cow.into_owned())
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Thanks for adding tests!
Would it be possible to add an error path testing to cover some of the errors possible from calling `decode_q()`  / `decode_word()` ?

Reply:
Added, and also fixed a case of no encoded-word.

Reply:
Thanks!